### PR TITLE
feat: prevent pagination token re-use using context

### PR DESCRIPTION
**Pull request by glb**
In a service with multiple resource types, or a service used by multiple users, it can be valuable to prevent malicious actors from using pagination tokens outside of the context in which they are intended to be used.

This change uses the `encryption_context` to bind pagination tokens to the authenticated user and the API path for which they were created. If someone else tries to use the token, or if a user tries to take a token from one resource and use it in a different resource, the decryption process will fail and the user request will be blocked.

In scenarios where the token encryption key is used across multiple API Gateway instances, it would be valuable to include the API Gateway ID in the context as well so that tokens cannot be taken across that boundary. If the token encryption key is not shared, the additional context of the API Gateway ID is unnecessary.

Thoughts?

### CODE REVIEW

**Comment by benkehoe** [comment]
There are two ways to approach this: using AAD, as you've done, or putting it in the encrypted data itself. Which do you think would be better for this situation?

File: src/common/pagination.py
Comment (benkehoe) [inline src/common/pagination.py]: The functions that know they are encrypting pagination tokens should take the desired context directly, e.g., user and path, rather than an arbitrary, untyped context.

File: src/common/pagination.py
Comment (benkehoe) [inline src/common/pagination.py]: Prefer `Optional[T]` to `Union[None, T]`; they are equivalent but `Optional` is a little clearer about intent. It's murkier when you already have a union, but people seem to prefer `Optional[Union[T1, T2]]` to `Union[None, T1, T2]` in [this poll I ran](https://twitter.com/ben11kehoe/status/1484557670463905797?s=20).

File: src/list_function.py
Comment (benkehoe) [inline src/list_function.py]: I'd wrap this `get_user(event)` function up in something like `common/event_utils.py`

**Review by benkehoe — CHANGES_REQUESTED** [review]
Add a section to `README.md` explaining why this is useful. Eventually all the architectural decisions are going to be documented.

File: src/common/pagination.py
Comment (glb) [reply src/common/pagination.py]: That's fair. In this case I'd suggest including the API Gateway API ID in the context as this deep in the utility there's no guarantee that the caller knows about the convention "don't share encryption material across API Gateway instances". Thoughts? 

Related: would you suggest creating an opinionated `PaginationEncryptionContext` class that can be instantiated using an event and has an `as_dict` method to satisfy the encryption SDK? What's the Python naming convention here -- I feel like `pagination.PaginationEncryptionContext` doesn't scan well, but `import EncryptionContext from common.pagination` loses meaning.

**Comment by glb** [comment]
> There are two ways to approach this: using AAD, as you've done, or putting it in the encrypted data itself. Which do you think would be better for this situation?

I think it's better to use AAD so that the decryption fails utterly when the context doesn't match and there's no opportunity for someone to forget to check the content. It also doesn't confuse concerns / lets you use whatever data you want in the token itself, naturally completely separate from the authentication mechanism without needing to invent an envelope format.

Thoughts?

File: src/common/pagination.py
Comment (benkehoe) [reply src/common/pagination.py]: Is it true, though, that we don't want to share encryption material across API GW instances? Are there situations where that might be justified?

I would say I'd prefer `encode_pagination_token(pagination_token, api_id, user_id, path)` to `encode_pagination_token(pagination_token, context)`

Comment (benkehoe) [reply src/common/pagination.py]: That's because these functions are defined in, and dedicated to, this service. If it was, instead, a general-purpose library for encoding pagination tokens (something that may be worth doing!) a flexible Context with helper functions to extract from an event, etc. would probably make sense.

**Comment by benkehoe** [comment]
> I think it's better to use AAD so that the decryption fails utterly when the context doesn't match

This makes sense, though the error needs to be handled separately but returned in the same way ("Invalid pagination token") as when some validation of the pagination token fails.

Another aspect is, AAD is not private, so we would be disclosing, for example, the API GW id if we include it in there.

File: src/common/pagination.py
Comment (glb) [reply src/common/pagination.py]: > Is it true, though, that we don't want to share encryption material across API GW instances? Are there situations where that might be justified?

Different teams work in different ways. My thought process here was that if you don't share encryption material, then key + user + path is likely sufficient to ensure that tokens can't be re-used in an unintended context. If you do share encryption material, then key + user + APIGW ID + path may be needed to be positive that you don't have collisions (though it's pretty unlikely that path would be the same across two APIGW's that share key material).

> I would say I'd prefer `encode_pagination_token(pagination_token, api_id, user_id, path)` to `encode_pagination_token(pagination_token, context)`

I've developed an allergy to >2 positional parameters of the same type in a row; while in this context we're likely to always use variables, I've seen too many calls that look like `fn("a", "b", "c", "d")` and struggled to understand what each parameter was for. I'd be far more comfortable with either a class that can extract from an event or named parameters. At the end of the day, though, it's your repository I'm writing a PR for. 🙂 

Thoughts on building a proto-Context here with the idea that it could be pulled out along with the rest of `pagination.py` to a general-purpose library if that became interesting?

Comment (benkehoe) [reply src/common/pagination.py]: > Different teams work in different ways.

Which would we rather recommend: "use different keys" or "use different context"? Like if we had to pick. I'm not sure but I don't think the answer is recommending "use different keys AND different context", but maybe it is.

> I've developed an allergy to >2 positional parameters of the same type in a row

This is absolutely correct, and I noticed in this review that I had not made these arguments keyword-only. I mean we should define it as something like `encode_pagination_token(pagination_token, *, path, user=None)`

> Thoughts on building a proto-Context here

I'd rather do it separately. One option is to build that separate library now and then have this project use it.

**Comment by glb** [comment]
> > I think it's better to use AAD so that the decryption fails utterly when the context doesn't match
> 
> This makes sense, though the error needs to be handled separately but returned in the same way ("Invalid pagination token") as when some validation of the pagination token fails.

My take on it was that the decryption error needs to be handled the same way as if the token had been tampered with. I didn't see any handling of a decryption exception so didn't add that.
 
> Another aspect is, AAD is not private, so we would be disclosing, for example, the API GW id if we include it in there.

We're a bit out of my depth here, could you help me understand this concern a bit better? I understand that [the encryption context will be logged in CloudTrail](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#encrypt_context), but I didn't think that it was included in the ciphertext.

**Comment by benkehoe** [comment]
It's a little strange, and I don't think there's an option to do otherwise, but the AAD is [included in the ciphertext blob](https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/message-format.html#header-structure) by the AWS Encryption SDK.

**Comment by benkehoe** [comment]
> I didn't see any handling of a decryption exception

[We handle errors in decryption here](https://github.com/benkehoe/serverless-example-project/blob/62231bf62cd1fcf5ff171be138e6ec4832fce00a/src/common/pagination.py#L144). We don't handle an invalid ExclusiveStartKey (that is, the parsed pagination token), because we assume we've gotten it from DDB. That's an argument in favor of keeping the context on the encryption side.

The other thing I realized is that there's an option to have unencrypted pagination tokens. In this case, if we use AAD for this context, that won't be included in those tokens. There's an argument both ways here; it's not authenticated so we shouldn't trust it anyway vs it's a safety feature in addition to a security feature, so having them in unencrypted tokens would be good too. I guess in theory we could wrap the unencrypted token so it has its own context, but that also feels weird.

**Comment by glb** [comment]
> It's a little strange, and I don't think there's an option to do otherwise, but the AAD is [included in the ciphertext blob](https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/message-format.html#header-structure) by the AWS Encryption SDK.

Well, that is ... interesting. Thank you for the pointer and unfortunately the reason for me to avoid using the encryption SDK. I am sure that there's a good reason for why they made that choice, but it seems really really undesirable to be unable to opt out of that behaviour. I get that some implementations need to store _part of_ the additional data in a sidecar, but always including all of it, wow. It makes me want to go spelunking for stuff that's been encrypted with the SDK and see what could get pulled out of the AAD content.

> [We handle errors in decryption here](https://github.com/benkehoe/serverless-example-project/blob/62231bf62cd1fcf5ff171be138e6ec4832fce00a/src/common/pagination.py#L144). We don't handle an invalid ExclusiveStartKey (that is, the parsed pagination token), because we assume we've gotten it from DDB. That's an argument in favor of keeping the context on the encryption side.

😳 my bad, apologies for not reading more carefully.

> I realized is that there's an option to have unencrypted pagination tokens.

I was treating unencrypted pagination tokens as the radioactive learning-only option and hoping that the to-be-filled-out README would have words to that effect. There's no protection against tampering with an unencrypted token. If this PR is "belt+suspenders", unencrypted pagination tokens (at least as implemented here -- I've seen implementations that are safe) are like not wearing pants.

File: src/common/pagination.py
Comment (glb) [reply src/common/pagination.py]: > Which would we rather recommend: "use different keys" or "use different context"? Like if we had to pick. I'm not sure but I don't think the answer is recommending "use different keys AND different context", but maybe it is.

My default would be "use different keys". The teams I work with deploy their keys as part of their service stack; each service has its own APIGW. It is completely and utterly coincidental (sincerely) that this means they wouldn't need the APIGW in the context.
 
> > I've developed an allergy to >2 positional parameters of the same type in a row
> 
> This is absolutely correct, and I noticed in this review that I had not made these arguments keyword-only. I mean we should define it as something like `encode_pagination_token(pagination_token, *, path, user=None)`
> > Thoughts on building a proto-Context here
> 
> I'd rather do it separately. One option is to build that separate library now and then have this project use it.

Whatever works for you. If you don't want to sidetrack, I'm happy to change this PR to use a bunch of keyword parameters.

**Comment by benkehoe** [comment]
> unencrypted pagination tokens as the radioactive learning-only option

I'm not sure on this point. As long as the following two conditions are met for a list operation:
* Irrelevant records are filtered
* There is no capability for returning some subset of the results, i.e., that relevant results should be hidden

I'm not sure there's a strong argument that security could be compromised if a user passes in an exclusive start key of their own making

**Comment by benkehoe** [comment]
> I'm not sure there's a strong argument that security could be compromised if a user passes in an exclusive start key of their own making

Actually wait, that's not entire true. So on the one hand, non-opaque pagination is definitely present in a lot of APIs, whether it's by counting the number of results, etc. But, with a single-table design, the LastEvaluatedKey could be a record that was filtered out of the returned results, and the key could contain information you wouldn't want to disclose. I still think the *integrity* of the key probably isn't an issue, but the confidentiality would be.

**Comment by glb** [comment]
> I still think the integrity of the key probably isn't an issue, but the confidentiality would be.

I expect that you're right on the integrity bit, at least in so far as DynamoDB will likely do The Right Thing™ if given tampered input. I tend to err on the side of "if the caller can't crack it open and mess with it, they can't build a dependency on it".

Getting back to your point, I think that we're agreed that unencrypted tokens have no integrity or confidentiality protection, so there's not a lot of point in adding protection against being re-used in other contexts. Yes?

**Comment by glb** [comment]
It has occurred to me that there's a pretty-standard practice for solving the "hey, I have some data that I want to check is the same at points A and B, but I don't particularly need/want anyone else to know what the data is" problem: Just Hash It™.

Instead of:

```
    pagination_token_encryption_context = {
        # Ensure that a token issued for one user can't be used by another user.
        # Customize this to use the appropriate identity field for your authorizer.
        "user": ((event.get("requestContext") or {}).get("identity") or {}).get("user") or "",
        # Ensure that a token issued for one resource type / path can't be used for
        # another resource type.
        "path": event.get("path") or "",
    }
```

I could use:

```
    h = hashlib.sha256()
    h.update(get_user(event))
    h.update(event.get("path") or "")
    h.update(get_api_id(event))

    pagination_token_context_digest = h.hexdigest()
```

and then pass the digest instead of tunnelling a bunch of parameters down to the bit that actually does the encryption. This also neatly avoids the problem of the function signature changing if someone wants to change the set of parameters used.

Thoughts?

**Comment by benkehoe** [comment]
Hashing keeps the information secret, but it doesn't change the fundamental mode of action: because the AAD is in the cipher text header, there's no way to make the decryption itself fail. So there's not a huge difference between putting a hashed API ID in the AAD or including it in the encrypted data itself, either way it needs to be extracted and compared against the value(s) for the current request. I'm checking with the AWS Encryption SDK team to see if they have any guidance on this.

**Comment by benkehoe** [comment]
The other way to do this would be key derivation, but the AWS Encryption SDK doesn't make that easy either.

**Comment by benkehoe** [comment]
Ok, I got to talk to some AWS Encryption SDK folks about this. Let's use AAD for this purpose. I think we've got two options: the context as a dict of key->hashed-value, or one-fixed-key->hash-of-all-the-values (or maybe hash-of-serialized-json-of-keys-and-values)

**Comment by glb** [comment]
Sounds good. I'm leaning towards one-fixed-key -> hash-of-all-the-values at this point, will see how if that still feels right once I've got it typed out.

**Comment by benkehoe** [comment]
I'd say the value you're hashing should still probably be `json.dumps({"api_id": get_api_id(event), ...}, sort_keys=True)`, so that there isn't concern about just concatenating strings with no separators.

The other thing I was thinking about was, doing a single hash bakes in the keys you're using, as opposed to allowing new keys to be added later. Though maybe it would just be like, `context_digest_1`, and then if you've got new keys to use you'd add `context_digest_2` for that hash.

**Comment by glb** [comment]
Added an attempt at addressing most of the comments above; still needs some documentation written but will do that and squash commits once we land on an approach 🙂

I believe that modifying the encryption context by adding more keys would be a breaking change (all pagination tokens with the old context wouldn't be decryptable with the new context), but it's entirely possible that the encryption SDK does something I wouldn't expect. Still need to do some testing to confirm my understanding.

File: src/list_function.py
Comment (benkehoe) [inline src/list_function.py]: I kind of think this should be inside the encode/decode functions, but it also seems kind of messy to do it that way

File: src/common/event_utils.py
Comment (benkehoe) [inline src/common/event_utils.py]: Fewer one liners in here. More like:
```python
if "requestContext" not in event or "identity" not in event["requestContext"]:
    return ""
return event["requestContext"]["identity"].get("user") or ""
```

File: src/common/event_utils.py
Comment (benkehoe) [inline src/common/event_utils.py]: pull all these values out into their own functions

File: src/common/event_utils.py
Comment (benkehoe) [inline src/common/event_utils.py]: This function should be in `pagination.py`. I also don't think you've actually defined `PaginationEncryptionContext` anywhere.

File: src/common/event_utils.py
Comment (benkehoe) [inline src/common/event_utils.py]: Let's use the key `context_digest_1`.

File: src/common/pagination.py
Comment (benkehoe) [inline src/common/pagination.py]: [`decrypt()`](https://aws-encryption-sdk-python.readthedocs.io/en/latest/generated/aws_encryption_sdk.html#aws_encryption_sdk.EncryptionSDKClient.decrypt) does not take a context. What we need is:
```python
if "context_digest_1" not in header.encryption_context:
    raise EncryptionContextException("Missing context_digest_1") # define EncryptionContextException above
if header.encryption_context["context_digest_1"] != encryption_context.context_digest_1:
    raise EncryptionContextException("Mismatch for context_digest_1")
```
actually the above should be wrapped in a method like `PaginationEncryptionContext.validate(encryption_context)`

File: src/common/event_utils.py
Comment (benkehoe) [reply src/common/event_utils.py]: I'd make `PaginationEncryptionContext` a [frozen dataclass](https://docs.python.org/3/library/dataclasses.html#frozen-instances)

**Comment by benkehoe** [comment]
>I believe that modifying the encryption context by adding more keys would be a breaking change (all pagination tokens with the old context wouldn't be decryptable with the new context), but it's entirely possible that the encryption SDK does something I wouldn't expect. Still need to do some testing to confirm my understanding.

This isn't how it works. Yes, if you have a different AAD in the decrypt operation it should fail, but the Encryption SDK packages the AAD with the ciphertext, and uses that packaged AAD in decrypting the data key via KMS. So unless you actually manually modify the ciphertext returned by `encrypt()`, `decrypt()` will always have the correct encryption context to decrypt the data key.

So instead, what we're doing is decrypting, and then comparing the encryption context returned by `decrypt()` (which it takes from the message header in the ciphertext) with the present context, and failing if the comparison reveals a mismatch. This is where if we add an additional set of context keys under a second `context_digest_2`, we could still allow old pagination tokens by only comparing the `context_digest_1` if it does not include `context_digest_2`.

Speaking of which, maybe we should add expiration as another context key.

**Comment by benkehoe** [comment]
If you want to give me commit rights to your repo, I could contribute some of those changes to this PR as well.

**Comment by benkehoe** [comment]
Updated, take a look and tell me what you think

File: src/list_function.py
Comment (glb) [inline src/list_function.py]: Off-topic: one of the things I like about Go is that it has a [built-in duration parser](https://pkg.go.dev/time#ParseDuration) so that a person doesn't need to calculate how many seconds they need if they want a 15-minute duration, for example (I wish I didn't have various increments burned into my brain). I saw a [neat implementation of something similar recently for Python](https://gist.github.com/benkehoe/5b03c308b038b29e42106f602e554010), you might be familiar with it. Thoughts on applicability here?

File: src/common/pagination.py
Comment (glb) [inline src/common/pagination.py]: nit: typo (throughout)

```suggestion
    DISABLE_PAGINATION_CONTEXT_STRICT_VALIDATION_ENV_KEY: ClassVar[
```

File: src/common/pagination.py
Comment (glb) [inline src/common/pagination.py]: I think you're passing in `last_evaluated_key` as `pagination_token`, right? 

```
        response["NextToken"] = encode_pagination_token(
             last_evaluated_key,
```

Which is taken directly from the response:

```
last_evaluated_key = response.get("LastEvaluatedKey")
```

Which (I think) [is defined as a dict](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb.html#DynamoDB.Table.scan).

Following through the code, I don't think that actually breaks, because the `context.package` takes `Any` as the token type, but seems like a broken link typing-wise here.

```suggestion
    pagination_token_data: Any,
```

File: src/common/pagination.py
Comment (glb) [inline src/common/pagination.py]: The `ascii` bits here surprised me, as my understanding is that JSON is canonically UTF-8. Today I learned that `json.dumps(..., ensure_ascii=True)` is a thing ([and the default behaviour](https://docs.python.org/3/library/json.html#character-encodings)). This feels like something you've been burned by in the past 🙂 Tucking this away for future pondering.

File: src/common/encryption.py
Comment (glb) [inline src/common/encryption.py]: Seems like `encryption_context` can be removed entirely now?

File: src/common/encryption.py
Comment (glb) [inline src/common/encryption.py]: Curious: what led you to choose `100` here?

File: CONTRIBUTORS.md
Comment (glb) [inline CONTRIBUTORS.md]: You've raised the bar so much here that this is an extremely generous attribution 🙂

**Review by glb — COMMENTED** [review]
One of your earlier comments was a request to add some detail to the `README`. Is that something you'd still like to see or do you think the additional comments in the code are sufficient?

Thank you for raising the bar so much on this PR! I've learned a lot from reading your changes. There are a couple of typos and what I think is a copy/paste error that I've highlighted in the comments.

File: CONTRIBUTORS.md
Comment (benkehoe) [reply CONTRIBUTORS.md]: Agree to disagree.

File: src/common/encryption.py
Comment (benkehoe) [reply src/common/encryption.py]: Well, I'm also now wondering if we should return the headers from the encrypt/decrypt operations

File: src/common/encryption.py
Comment (benkehoe) [reply src/common/encryption.py]: I did not really find any guidance on it. This example uses 100? https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/sample-cache-example-code.html

File: src/common/pagination.py
Comment (benkehoe) [reply src/common/pagination.py]: I guess we could do it either way.

File: src/list_function.py
Comment (benkehoe) [reply src/list_function.py]: Sure, if you wanted to pull that code in, we could use it here.

File: src/common/pagination.py
Comment (benkehoe) [inline src/common/pagination.py]: We should call this `validated_pagination_token_data` I think, given that the input to this function is called `pagination_token` already

File: src/common/pagination.py
Comment (benkehoe) [reply src/common/pagination.py]: yes, and it looks like the call to `_encode...` similarly has the parameter misnamed

**Review by benkehoe — CHANGES_REQUESTED** [review]

File: src/common/pagination.py
Comment (glb) [reply src/common/pagination.py]: Looks like you've done this, resolving.

File: src/common/encryption.py
Comment (glb) [reply src/common/encryption.py]: Not sure. What are you thinking we'd do with them?

**Comment by glb** [comment]
Aging this out. I doubt I'll be able to get back to it. Thanks for the patient conversation.

**Pull request closed without merging**
